Simplify write failure handling

areek opened the pull request:
Currently, any write (e.g. `index`, `delete`) operation failure can be categorized as:
- request failure (e.g. analysis, parsing error, version conflict) 
- transient operation failure (e.g. due to shard initializing, relocation)
- environment failure (e.g. out of disk, corruption, lucene tragic event)

The main motivation of the PR is to handle these failure types appropriately for a 
write request. Each failure type needs to be handled differently:
- request failure (being request specific) should be replicated and then failed
- transient failure should be retried (eventually succeeding)
- environment failure (persistent primary shard failure) should fail the request 
  immediately.

Currently, transient operation failures are retried in replication action but no distinction
is made between request and environment failures, both fails write request immediately. 

In this PR, we distinguish between request and environment failures for a write operation. 
In case of environment failures, the exception is bubbled up failing the request and in case 
of request failures, the exception is captured and replication continues (we ignore performing 
on replicas when such failures occur in primary). Transient operation failures are bubbled up 
to be retried by the replication operation, as before. 

Note: https://github.com/elastic/elasticsearch/issues/20109 simplifies bulk execution code, which should clean up error handling for shard bulk requests.

areek commented:
After discussions with @bleskes, I changed the scope of this PR. Now, the PR focuses on making primary write operation failures a valid write result, so replication operation can handle them explicitly. We can add operation failure replication, as needed in the `feature/seq_no` branch

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -152,31 +151,30 @@ protected boolean resolveIndex() {
+        if (result.operationFailed()) {
             // rethrow the failure if we are going to retry on primary and let parent failure to handle it
-            if (retryPrimaryException(e)) {
+            if (retryPrimaryException(result.getFailure())) {
```
nite: maybe just put this in a variable?

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -152,31 +151,30 @@ protected boolean resolveIndex() {
                     applyVersion(request.items()[j], preVersions[j], preVersionTypes[j]);
                 }
-                throw (ElasticsearchException) e;
+                throw (ElasticsearchException) result.getFailure();
```
this smells - why do we explicitly throw an error?

areek replied:
this was left over, thanks for pointing it out!

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -193,32 +191,31 @@ protected boolean resolveIndex() {
             // if its a conflict failure, and we already executed the request on a primary (and we execute it
             // again, due to primary relocation and only processing up to N bulk items when the shard gets closed)
             // then just use the response we got from the successful execution
-            if (item.getPrimaryResponse() != null && isConflictException(e)) {
```
why did you change the isConflictException() check to an instance of?

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -403,29 +400,23 @@ private UpdateResult shardUpdateOperation(IndexMetaData metaData, BulkShardReque
-                    WriteResult result = shardIndexOperation(bulkShardRequest, indexRequest, metaData, indexShard, false);
+                WriteResult result = shardIndexOperation(bulkShardRequest, indexRequest, metaData, indexShard, false);
+                if (result.operationFailed()) {
+                    return new UpdateResult(translate, indexRequest,
```
same question about change in unwrapping

bleskes commented on core/src/main/java/org/elasticsearch/action/delete/TransportDeleteAction.java:
```diff
@@ -130,18 +130,24 @@ protected Location onReplicaShard(DeleteRequest request, IndexShard indexShard)
     public static Engine.Delete executeDeleteRequestOnReplica(DeleteRequest request, IndexShard indexShard) {
         Engine.Delete delete = indexShard.prepareDeleteOnReplica(request.type(), request.id(), request.version(), request.versionType());
         indexShard.delete(delete);
+        if (delete.getFailure() != null) {
```
hmm... I think we need to change replica to return a write result as well.. this is to ugly :( ( and I think we misunderstood each other on "not doing anything with the failures on replicas for now.. "

areek replied:
aah, now we communicate replica failures via action listeners in replication operation.

bleskes commented on core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java:
```diff
@@ -140,31 +139,39 @@ protected IndexResponse newResponseInstance() {
+        final Engine.Index operation;
+        try {
+            operation = indexShard.prepareIndexOnReplica(sourceToParse, request.version(), request.versionType());
+        } catch (Exception e) {
```
what do we throw anything here? is this when the parsing doesn't match the mapping?

areek replied:
we throw error while parsing index request source here, currently we bubble up the exception as this is not engine level exception (though it is operation specific) .

bleskes commented on core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java:
```diff
@@ -175,29 +182,40 @@ protected Location onReplicaShard(IndexRequest request, IndexShard indexShard) {
-            mappingUpdatedAction.updateMappingOnMaster(shardId.getIndex(), request.type(), update);
-            operation = prepareIndexOperationOnPrimary(request, indexShard);
+            try {
+                mappingUpdatedAction.updateMappingOnMaster(shardId.getIndex(), request.type(), update);
```
do we expect this method to ever throw an exception?

areek replied:
Yes, due to timing out for example. Now we bubble the exception up to be handled in the replication action instead of capturing it

bleskes commented on core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java:
```diff
@@ -175,29 +182,40 @@ protected Location onReplicaShard(IndexRequest request, IndexShard indexShard) {
+            try {
+                mappingUpdatedAction.updateMappingOnMaster(shardId.getIndex(), request.type(), update);
+                operation = prepareIndexOperationOnPrimary(request, indexShard);
+            } catch (Exception e) {
```
I'm on the fence regarding these - originally I was thinking we should limit the failure handling to things that happened on the engine level. Not saying this is bad but I want to hear why you decided to add this.

areek replied:
Now we let these failures bubble up to be retried by the replication operation. Initially, I thought it would be a good idea to handle these failures explicitly by inspecting the write result at the replication operation layer, now I see it makes more sense to let replication operation deal with these as is and only capture the operation-level failures to be dwelt explicitly

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:
```diff
@@ -367,7 +369,7 @@ public RetryOnPrimaryException(StreamInput in) throws IOException {
     public interface PrimaryResult<R extends ReplicationRequest<R>> {
 
-        R replicaRequest();
+        @Nullable R replicaRequest();
```
can we document what null means?

areek replied:
added documentation

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -152,32 +150,31 @@ protected boolean resolveIndex() {
             }
+            return null;
         }
+        location = locationToSync(location, result.getLocation());
```
can we put this first under result.operationFailed() == false? I think it will be easier to read for most people (they want to see the "normal" flow and it's shorter? als we don't run the risk of forgetting a return statement because the compiler will tell us

areek replied:
agreed, I changed it to `result.success()` as suggested

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -152,32 +150,31 @@ protected boolean resolveIndex() {
+        WriteResult<IndexResponse> result = shardIndexOperation(request, indexRequest, metaData, indexShard, true);
+        if (result.operationFailed()) {
+            Exception failure = result.getFailure();
+            if (retryPrimaryException(failure)) {
```
these are not an operation level issue and should cause the entire request to fail and be retried

areek replied:
Now primary retry exceptions are bubbled up

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -194,32 +191,31 @@ protected boolean resolveIndex() {
+        final WriteResult<DeleteResponse> writeResult = TransportDeleteAction.executeDeleteRequestOnPrimary(deleteRequest, indexShard);
+        if (writeResult.operationFailed()) {
+            Exception failure = writeResult.getFailure();
+            if (retryPrimaryException(failure)) {
```
same comments as the index case

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -229,12 +225,7 @@ protected boolean resolveIndex() {
-                updateResult = new UpdateResult(null, null, false, t, null);
-            }
+            UpdateResult updateResult = shardUpdateOperation(metaData, request, updateRequest, indexShard);
             if (updateResult.success()) {
```
since this one already uses the `.success()` method, maybe do the same on WriteResult?

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -346,14 +348,14 @@ private boolean checkVersionConflict(
-                        op.versionType().explainConflictForWrites(currentVersion, expectedVersion, deleted));
+                op.setFailure(new VersionConflictEngineException(shardId, op.type(), op.id(),
+                        op.versionType().explainConflictForWrites(currentVersion, expectedVersion, deleted)));
+            } else {
```
I think we can remove this empty thing?

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -346,14 +348,14 @@ private boolean checkVersionConflict(
             }
+            return true;
         }
         return false;
```
can this be under else {}?

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -394,53 +396,68 @@ private long checkDeletedAndGCed(VersionValue versionValue) {
-                currentVersion = loadCurrentVersionFromIndex(index.uid());
+                try {
+                    currentVersion = loadCurrentVersionFromIndex(index.uid());
+                } catch (IOException e) {
```
I think it will be easier /cleaner to deal with all the exceptions on the index() method..

areek replied:
Agreed, now the index operation failure is only set on the `index` and `delete` method

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -515,30 +515,25 @@ private IndexShardState changeState(IndexShardState newState, String reason) {
      * Index a document and return whether it was created, as opposed to just
```
this comment is now wrong

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/TranslogRecoveryPerformer.java:
```diff
@@ -159,6 +159,9 @@ private void performRecoveryOperation(Engine engine, Translog.Operation operatio
                     }
                     index(engine, engineIndex);
+                    if (engineIndex.getFailure() != null) {
+                        throw engineIndex.getFailure();
```
I think we can return the engine.operation here and then check on the caller side if there were errors we care about. I think we won't care about most of them.

jasontedor commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -394,53 +396,68 @@ private long checkDeletedAndGCed(VersionValue versionValue) {
-                maybeFailEngine("index", e);
-            } catch (Exception inner) {
-                e.addSuppressed(inner);
+        } catch (OutOfMemoryError | IllegalStateException t) {
```
This will swallow and `OutOfMemoryError` which we definitely do not want to do and it's questionable to me that we catch it all?

areek replied:
I changed it not to catch the OOM, this was not intentional, must have crept in while updating the pr to master

areek replied:
@jasontedor in `Engine#maybeFailEngine` we still check for an OOM exception to fail the engine, is this left over?

jasontedor replied:
Yeah, I do not think that we should be catching `OutOfMemoryException` or try to fail gracefully upon such here. It would be a disaster if during the handling of that exception (i.e., subsequently trying to fail the engine), another exception is thrown, and the original `OutOfMemoryException` is lost and never bubbles up to cause the VM to tear itself down.

bleskes replied:
I agree this is not inline with our current trend, but it’s a different change imo. Maybe deal with it (and the discussion that will follow) on a different PR?

> On 01 Aug 2016, at 12:19, Jason Tedor <notifications@github.com> wrote:
> 
> In core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
> 
> >                  }
> >              }
> > -        } catch (IllegalStateException | IOException e) {
> > -            try {
> > -                maybeFailEngine("index", e);
> > -            } catch (Exception inner) {
> > -                e.addSuppressed(inner);
> > +        } catch (OutOfMemoryError | IllegalStateException t) {
> 
> Yeah, I do not think that we should be catching OutOfMemoryException or try to fail gracefully upon such here. It would be a disaster if during the handling of that exception (i.e., subsequently trying to fail the engine), another exception is thrown, and the original OutOfMemoryException is lost and never bubbles up to cause the VM to tear itself down.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jasontedor replied:
@bleskes The conversation belongs here because we are not catching it in master (I already removed it in #19231) and I noticed a regression slip in here. :smile:

jasontedor replied:
Thanks for noticing this @areek, I opened #19881.

jasontedor commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -450,56 +467,53 @@ private long updateVersion(Engine.Operation op, long currentVersion, long expect
-                e.addSuppressed(inner);
-            }
-            throw new DeleteFailedEngineException(shardId, delete, e);
+        } catch (OutOfMemoryError | IllegalStateException t) {
```
Same comment here. This will also swallow an `OutOfMemoryError` which we definitely do not want to do and it's questionable to me that we catch it all?

areek commented:
@bleskes I updated the PR to only capture operation level write failures in write results for primary and replica operations that do not fail the engine and are not dwelt with (retried) by the replication action.  Could you take a look?

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -148,35 +145,41 @@ protected boolean resolveIndex() {
+                        new BulkItemResponse.Failure(request.index(), indexRequest.type(), indexRequest.id(), result.getFailure())));
+                }
+            }
         } catch (Exception e) {
```
I wonder if this fits in the new philosophy - all "valid" item level failures should not be thrown but rather be channelled into result.getFailure(), so everything that comes in here should not be shoved into  an item level failure, but rather bubbled up. Since we want to have special logic in the retryPrimaryException , why not just catch ElasticsearchException? we can also addes asserts to retryPrimaryException to make sure it only returns true if it's paramater is  subclass of ElasticsearchException?

areek replied:
I agree this fits the new philosophy, now we only bubble up retryPrimaryExceptions to be handled by replication operation and only result failures are set as bulk item level failures

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -148,35 +145,41 @@ protected boolean resolveIndex() {
     }
 
-    private Translog.Location index(IndexMetaData metaData, BulkShardRequest request, IndexShard indexShard, long[] preVersions, VersionType[] preVersionTypes, Translog.Location location, int requestIndex, BulkItemRequest item) {
+    private Translog.Location index(IndexMetaData metaData, BulkShardRequest request, IndexShard indexShard, long[] preVersions, VersionType[] preVersionTypes, Translog.Location location, int requestIndex, BulkItemRequest item) throws Exception {
```
I know it's not your fault, but can we add some java docs, mostly around the return value relationship to the location parameter? (which better named currentLocationToSync or so)

areek replied:
I will address this in https://github.com/elastic/elasticsearch/issues/20109

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -196,34 +199,39 @@ protected boolean resolveIndex() {
+            if (writeResult.success()) {
+                DeleteResponse deleteResponse = writeResult.getResponse();
+                location = locationToSync(location, writeResult.getLocation());
+                setResponse(item, new BulkItemResponse(item.id(), OP_TYPE_DELETE, deleteResponse));
```
I think this OP_TYPE_DELETE is not needed now that we have `Operation` (which was renamed to `Result`) on over DocWriteResponse. I think we can fold all this success/failure logic to a shard method and call it from both the indexing path and delete path? will remove a lot of boiler plate

areek replied:
addressed in #20109

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -196,34 +199,39 @@ protected boolean resolveIndex() {
                     applyVersion(request.items()[j], preVersions[j], preVersionTypes[j]);
                 }
-                throw (ElasticsearchException) e;
+                throw e;
```
same comment as above, but out of curiosity - why doesn't this force one to add throws Exception to the method signature? it's smart enough to detect that everything that may bubble up is a runtime exception?

areek replied:
this doesn't force to add a throws Exception because `TransportDeleteAction#executeDeleteRequestOnPrimary` method didn't explicitly specify it throws Exception in the method signature.

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -405,27 +411,25 @@ private UpdateResult shardUpdateOperation(IndexMetaData metaData, BulkShardReque
                     }
-                    return new UpdateResult(translate, indexRequest, retry, cause, null);
+                } catch (Exception e) {
+                    return new UpdateResult(translate, indexRequest, false, ExceptionsHelper.unwrapCause(e), null);
```
what failures does this catch? do we want to mark them as doc level errors?

areek replied:
Before this would catch parsing errors while indexing, but now we mark them as doc level errors and communicate it through `PrimaryOperationResult`

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -405,27 +411,25 @@ private UpdateResult shardUpdateOperation(IndexMetaData metaData, BulkShardReque
+                        return new UpdateResult(translate, deleteRequest, isConflictException(deleteResult.getFailure()), ExceptionsHelper.unwrapCause(deleteResult.getFailure()), null);
                     }
-                    return new UpdateResult(translate, deleteRequest, retry, cause, null);
+                } catch (Exception e) {
```
same question.

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -447,8 +451,10 @@ protected Location onReplicaShard(BulkShardRequest request, IndexShard indexShar
+                    if (result.success()) {
+                        location = locationToSync(location, result.getLocation());
+                    }
                 } catch (Exception e) {
```
can you double check we need to protect against ignoreReplicaExceptions? I think we do that high up the stack.

areek replied:
We do check for ignoreReplicaExceptions at `ReplicaOperation`

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -473,7 +480,7 @@ protected Location onReplicaShard(BulkShardRequest request, IndexShard indexShar
             }
         }
-        return location;
+        return new WriteResult<>(null, location);
```
this null feels wrong. Can we have a ReplicaWriteResult or something else that can handle the replica case? Maybe have WriteResult not require a response and have a PrimaryWriteResult inherit from and require a response?

areek replied:
changed, now we have `ReplicaWriteResult` for handling the replica case, so no need for passing null for the response.

bleskes commented on core/src/main/java/org/elasticsearch/action/delete/TransportDeleteAction.java:
```diff
@@ -123,13 +122,16 @@ protected DeleteResponse newResponseInstance() {
         indexShard.delete(delete);
+        if (delete.hasFailure()) {
+            return new WriteResult<>(delete.getFailure());
+        }
```
can we have an else (channeling Simon :))

bleskes commented on core/src/main/java/org/elasticsearch/action/delete/TransportDeleteAction.java:
```diff
@@ -139,9 +141,12 @@ protected Location onReplicaShard(DeleteRequest request, IndexShard indexShard)
-        return delete;
+        if (delete.hasFailure()) {
+            return new WriteResult<>(delete.getFailure());
+        }
```
same

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:
```diff
@@ -367,7 +370,11 @@ public RetryOnPrimaryException(StreamInput in) throws IOException {
-        R replicaRequest();
+        /**
+         * Returns the replication request when the primary operation
+         * is successful otherwise returns <code>null</code>
```
the description is wrong - we don't know if things were successful or not (and in fact with seq # we will replicate failures). I think we should just say "returns null if no operation needs to be sent to a replica (for example when the operation failed on the primary due to a parsing exception)"

bleskes replied:
Also, with this offical way of saying "no need to replicate", I think we can remove the `executeOnReplicas` paramter to the `ReplicationOperation`. This can be done as a follow up. Can you add this to your list? :)

areek replied:
> I think we can remove the executeOnReplicas paramter to the ReplicationOperation

I will work on this in a subsequent PR.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:
```diff
@@ -131,6 +153,15 @@ public WritePrimaryResult(Request request, Response finalResponse,
+             * We call this before replication because this might wait for a refresh and that can take a while. This way we wait for the
+             * refresh in parallel on the primary and on the replica.
+             */
+            postWriteActions(indexShard, request, null, this, logger);
```
why do we need this when failing and not replicating?

Also - unrelated but will make this easier to understand. Can we make TransportWriteAction use the same distinction between primary and replica request on it's types?

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:
```diff
@@ -131,6 +153,15 @@ public WritePrimaryResult(Request request, Response finalResponse,
             postWriteActions(indexShard, request, location, this, logger);
         }
 
+        public WritePrimaryResult(Request request, Exception finalFailure, IndexShard indexShard) {
```
we should mark request as nullable no?

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:
```diff
@@ -161,10 +194,15 @@ public synchronized void respondAfterAsyncAction(boolean forcedRefresh) {
         }
 
+        public WriteReplicaResult(ReplicatedWriteRequest<?> request, Exception failure, IndexShard indexShard) {
+            this.failure = failure;
```
similar comments about nullable request and making a destinction in types.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -394,30 +396,34 @@ private long checkDeletedAndGCed(VersionValue versionValue) {
             try {
-                maybeFailEngine("index", e);
+                if (maybeFailEngine("index", e) == false) {
+                    // we set operation specific engine failure that doesn't
```
OK - I spent an hour researching this and this is the right thing to do, but we need way more explaination as to why :)

Here is a suggestion:

```
When indexing a document into Lucene, Lucene distinguishes between environment related errors (like out of disk space) and document specific errors (like analysis chain problems) by setting the IndexWriter.getTragicEvent() value for the former. maybeFailEngine checks for these kind of errors and returns true if that is the case. We use that information to indicate a document level failure and set the result on index.setFailure
```

bleskes replied:
one more issue I have with this - if we go down this route , we need to catch *any* exception , not just the above frew.

areek replied:
Thanks @bleskes for the suggestion. I have added in the suggestion and now we catch any exception

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -352,7 +354,7 @@ private boolean checkVersionConflict(
             } else {
                 // fatal version conflict
                 throw new VersionConflictEngineException(shardId, op.type(), op.id(),
```
can we not throw an exception but set it on the operation and return true?

bleskes replied:
also java doc all the tricky things ™️

bleskes replied:
and if we do so, let's remove VersionConflictEngineException from the out catch clause?

areek replied:
now we set version conflict exception as operation failure

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -450,56 +454,55 @@ private long updateVersion(Engine.Operation op, long currentVersion, long expect
+                if (maybeFailEngine("delete", e) == false) {
+                    // we set operation specific engine failure that doesn't
+                    // fail the engine instead of bubbling them up
+                    delete.setFailure(e);
```
same comments as the other one.

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -512,33 +512,32 @@ private IndexShardState changeState(IndexShardState newState, String reason) {
-            created = engine.index(index);
+            engine.index(index);
             index.endTime(System.nanoTime());
+            if (index.hasFailure()) {
```
can we do this out of the try catch? now an exception in postIndex will cause it to be called twice no?

areek replied:
Thanks for the suggestion! changed

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -572,18 +571,21 @@ public void delete(Engine.Delete delete) {
             engine.delete(delete);
             delete.endTime(System.nanoTime());
+            if (delete.hasFailure()) {
+                indexingOperationListeners.postDelete(delete, delete.getFailure());
```
same comment as before

bleskes commented:
Thx @areek . This is getting in good shape. I left some more suggestions. It's probably good to revisit the PR title and description. It has deviated a bit from our current thinking.

bleskes commented:
one more thing - master moved slightly further in a way that might be useful here. I suggest you merge master into your PR branch.

areek commented:
Thanks @bleskes for the review! sorry for the lateness here. I have updated the PR and the description, would love to get your feedback on this.

bleskes commented on core/src/main/java/org/elasticsearch/ElasticsearchException.java:
```diff
@@ -486,8 +487,6 @@ public String toString() {
                 org.elasticsearch.index.shard.TranslogRecoveryPerformer.BatchOperationException::new, 26),
         SNAPSHOT_CREATION_EXCEPTION(org.elasticsearch.snapshots.SnapshotCreationException.class,
                 org.elasticsearch.snapshots.SnapshotCreationException::new, 27),
-        DELETE_FAILED_ENGINE_EXCEPTION(org.elasticsearch.index.engine.DeleteFailedEngineException.class,
```
can leave a comment as to where ordinal 28 went?

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -127,10 +124,10 @@ protected boolean resolveIndex() {
     }
 
-    private Translog.Location handleItem(IndexMetaData metaData, BulkShardRequest request, IndexShard indexShard, long[] preVersions, VersionType[] preVersionTypes, Translog.Location location, int requestIndex, BulkItemRequest item) {
+    private Translog.Location handleItem(IndexMetaData metaData, BulkShardRequest request, IndexShard indexShard, long[] preVersions, VersionType[] preVersionTypes, Translog.Location location, int requestIndex, BulkItemRequest item) throws Exception {
```
why do we need the extra "throws exception"?  If we the intention is documentation, I think we're better off with java docs?

areek commented:
@bleskes Thanks for the feedback! I rewrote the PR against latest master (and updated PR description). Now instead of having a generic `execute` method for write operations in the engine level, I added a generic `execute` method in `IndexShard`, keeping the changes to `InternalEngine` minimal. Would love to get a review on this :)

s1monw commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -397,7 +397,7 @@ private long checkDeletedAndGCed(VersionValue versionValue) {
 
     @Override
-    public void index(Index index) {
+    public void index(Index index) throws OperationFailedEngineException {
```
no need to declare `OperationFailedEngineException` it's unchecked

areek replied:
I removed the declaration, thanks @s1monw

s1monw commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -408,13 +408,31 @@ public void index(Index index) {
+     * and set the error in operation.setFailure. In case of environment related errors, the failure
+     * is bubbled up
+     */
+    private void handleOperationFailure(final Operation operation, final Exception e) throws OperationFailedEngineException {
```
I am not sure if we should do this. We are basically catching everything and wrap it in an unchecked exception. That's not a good practice IMO. Why do we need to wrap this, can't we transport it by adding a suppressed exception or something like this. I am worried we are hiding the right exceptions...

areek replied:
I restructured the method to clarify what we are doing here. We only wrap failures that fail the engine for serialization. Any request/document specific failures are communicated to the transport via `Operation#setFailure` and `Operation#getFailure`.

bleskes commented:
I did a first sweep and I like the simplification of having `IndexShard` work on the operation level. I will do another careful sweep tomorrow and I want to think some more about the higher level replication logic. Also, if we end up merging the exception types - we need a BWC layer :)

areek commented:
Thanks @bleskes for your feedback, I updated the PR simplifying `TransportWriteAction` by decoupling write operations from write responses, though this added complexity to `TransportShardBulkAction`. I will update the PR with simple unit tests for transporting write failures and work on the BWC for merging the exception types.

areek commented:
@bleskes I have updated the PR, making the index/delete operations immutable by introducing a operation result class, among other things we discussed. Would be awesome if you could do a review.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -517,59 +530,62 @@ private void innerIndex(Index index) throws IOException {
-                // document does not exists, we can optimize for create
-                index(index, indexWriter);
+                // skip index operation because of version conflict on recovery
+                return new IndexResult(expectedVersion, false, index.estimatedSizeInBytes());
```
also , this should be frozen, which kind of makes my point :)

bleskes replied:
PS: for **another** change - I don't like the inconsistencies in how we handle version conflicts during recovery and how we handle them for replicas. For the first we ignore here. For the second we ignore the replication action universe. We should have one way for both.

areek replied:
I changed it to have one return statement with a final variable and calling freeze on it as the last operation. thanks for pointing it out.

I agree we should remove the inconsistency in handling version conflict during recovery and for replicas. I will open a separate PR for it

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -517,59 +530,62 @@ private void innerIndex(Index index) throws IOException {
+        } catch (Exception e) {
+            Exception documentFailure = extractDocumentFailure(delete, e);
+            result = new DeleteResult(documentFailure, delete.version(),
+                    delete.estimatedSizeInBytes());
```
freeze?

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -596,19 +615,31 @@ private void innerDelete(Delete delete) throws IOException {
-            maybeAddToTranslog(delete, updatedVersion, Translog.Delete::new, DeleteVersionValue::new);
+            if (checkVersionConflict(delete, currentVersion, expectedVersion, deleted)) {
+                // skip executing delete because of version conflict on recovery
+                return new DeleteResult(expectedVersion, true,
```
weird line wrap + the same request as before - one return statement (with a freeze)

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -536,26 +536,31 @@ private IndexShardState changeState(IndexShardState newState, String reason) {
-    private void index(Engine engine, Engine.Index index) {
+    private Engine.IndexResult index(final Engine engine, final Engine.Index index) {
         active.set(true);
-        index = indexingOperationListeners.preIndex(index);
```
I agree with this, but it's a different change and is not related to this one? shall we do it another PR?

areek replied:
sure, I reverted the change

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -536,26 +536,31 @@ private IndexShardState changeState(IndexShardState newState, String reason) {
             throw e;
         }
-        indexingOperationListeners.postIndex(index, index.isCreated());
+        if (result.hasFailure()) {
```
I'm wondering if should make the distinction here and not just call postIndex(index,result) where the result is a failure

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1117,7 +1126,8 @@ private void ensureWriteAllowed(Engine.Operation op) throws IllegalIndexShardSta
         if (shardRouting.primary() == false) {
-            throw new IllegalStateException("shard is not a primary " + shardRouting);
+            // TODO throw a more appropriate exception
+            throw new ShardNotFoundException(shardRouting.shardId(), "shard is not a primary anymore");
```
why did you changed it? I think it's good to keep it consistent with verifyReplicationTarget and not use an exception that is covered by isShardNotAvailableException. These days, with aId checks and all, end up here by mistake is something we shouldn't swallow.

bleskes commented on core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java:
```diff
@@ -718,7 +718,7 @@ protected ReplicaResult shardOperationOnReplica(Request request, IndexShard repl
                     throw new ElasticsearchException("simulated");
                 }
-                return new ReplicaResult();
+                return new ReplicaResult(null);
```
I expected to see some test that shows we respond with a failure if the (Primary|Replica)Result has a failure but I can't find it. Is there one?

areek replied:
I added tests for this in TransportWriteActionTests.

bleskes commented on core/src/test/java/org/elasticsearch/action/support/replication/TransportWriteActionTests.java:
```diff
@@ -128,27 +127,27 @@ public void testReplicaWaitForRefresh() throws Exception {
     }
 
-    private class TestAction extends TransportWriteAction<TestRequest, TestResponse> {
+    private class TestAction extends TransportWriteAction<TestRequest, TestRequest, TestResponse> {
```
same question about failure testing - where is it?

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:
```diff
@@ -135,6 +134,7 @@
 import static java.util.Collections.emptyMap;
 import static org.elasticsearch.index.engine.Engine.Operation.Origin.PRIMARY;
 import static org.elasticsearch.index.engine.Engine.Operation.Origin.REPLICA;
+import static org.hamcrest.CoreMatchers.instanceOf;
```
commenting here because it's the first line - can we have a test that mock lucene's addDocuments to throw an non-tragic exception and see that is caught as a document one?

areek replied:
I added unit test for `checkIfDocumentFailureOrThrow` and test mocking non-tragic exception from index writer to test failure handling when executing write/delete operations.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -277,9 +278,118 @@ public Condition newCondition() {
+            return version;
+        }
+
+        public Translog.Location getLocation() {
```
I think getTranslogLocation is clearer name?

areek replied:
changed

areek commented on core/src/main/java/org/elasticsearch/action/admin/indices/flush/TransportShardFlushAction.java:
```diff
@@ -54,14 +54,14 @@ protected ReplicationResponse newResponseInstance() {
         replica.flush(request.getRequest());
         logger.trace("{} flush request executed on replica", replica.shardId());
-        return new ReplicaResult();
+        return new ReplicaResult(null);
```
done

areek commented on core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java:
```diff
@@ -140,65 +140,90 @@ protected IndexResponse newResponseInstance() {
+        final Engine.Index operation;
+        try {
+            operation = replica.prepareIndexOnReplica(sourceToParse, request.version(), request.versionType(), request.getAutoGeneratedTimestamp(), request.isRetry());
+        } catch (MapperParsingException | IllegalArgumentException e) {
```
we can get IAE when merging dynamic mapping update because of incompatible data type for an already seen field (see FieldMapper#346)

jpountz replied:
It should never happen that we get this specific exception on the replica if we did not get it on the primary before as it would mean that two shards have a different type for a given field even though types cannot be changed once a field is added.

bleskes replied:
@jpountz thx, but also on the primary, it would be great to know that parsing issues bubble up as MapperParsingException-s

areek replied:
@jpountz IMO we should throw a `MapperException` instead of IAE in case of conflicts in merging dynamic mapping update. then we can catch only `MapperException` while preparing index operation (I can do this as a followup, if this sounds good). 
@bleskes I already removed catching the `IAE` when preparing index operation on the replica.

areek commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -118,30 +117,85 @@ protected boolean resolveIndex() {
             BulkItemRequest item = request.items()[requestIndex];
-            // add the response
-            setResponse(item, new BulkItemResponse(item.id(), opType, writeResult.getResponse()));
+            if (operationResult == null // in case of a noop update operation
```
thanks, it is easier to follow

areek commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -118,30 +117,85 @@ protected boolean resolveIndex() {
+            }
+            assert item.getPrimaryResponse() != null;
+            assert preVersionTypes[requestIndex] != null;
+            if (item.getPrimaryResponse().isFailed()
```
done

areek commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -151,147 +205,159 @@ protected boolean resolveIndex() {
+                throw e;
             }
+            // TODO: maybe this assert is too strict, we can still get environment failures while executing write operations
+            assert false : "unexpected exception: " + e.getMessage() + " class:" + e.getClass().getSimpleName();
```
Now, we just throw an ElasticsearchException with a shardID as header instead of `OperationFailedEngineException` when we encounter a persistent engine failure. In bulk execution, we bubble it up (similar to retryOnPrimary exceptions)

areek commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -151,147 +205,159 @@ protected boolean resolveIndex() {
+                if (updateOperationResult.hasFailure() == false) {
+                    // enrich update response and
+                    // set translated update (index/delete) request for replica execution in bulk items
+                    switch (updateOperationResult.getOperationType()) {
```
agreed!

areek commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -151,147 +205,159 @@ protected boolean resolveIndex() {
+                                        indexResponse.getVersion(), sourceAndContent.v2(), sourceAndContent.v1(), indexSourceAsBytes));
+                            }
+                            // replace the update request to the translated index request to execute on the replica.
+                            request.items()[requestIndex] = new BulkItemRequest(request.items()[requestIndex].id(), updateIndexRequest);
```
makes sense, I added `replicaRequest` to `UpdateResultsHolder`

areek commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -151,147 +205,159 @@ protected boolean resolveIndex() {
+                            request.items()[requestIndex] = new BulkItemRequest(request.items()[requestIndex].id(), updateDeleteRequest);
+                            break;
+                    }
+                } else {
```
changed

areek commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -386,45 +387,55 @@ private long checkDeletedAndGCed(VersionValue versionValue) {
+        if (isDocumentFailure) {
+            return failure;
+        } else {
+            throw new OperationFailedEngineException(shardId, operation.operationType().getLowercase(),
```
This makes a lot of sense to me :) I changed it to throw an ElasticsearchException with a shard header.

areek commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -206,8 +206,13 @@ protected TransportRequestOptions transportOptions() {
+        boolean retry = e.getClass() == ReplicationOperation.RetryOnPrimaryException.class
+                || TransportActions.isShardNotAvailableException(e);
+        if (retry) {
+            assert e instanceof ElasticsearchException
```
This was to ensure all retryPrimaryExceptions are indeed instance of ElasticsearchException, I added this to make sure none of the retryPrimaryExceptions were silently shallowed at TransportShardBulkAction. I reverted this change. see https://github.com/elastic/elasticsearch/pull/19105#r73494534 for previous discussion on this.

areek commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -151,147 +205,159 @@ protected boolean resolveIndex() {
+                        // check if any transient write operation failures should be bubbled up
+                        Exception failure = operationResult.getFailure();
+                        if (!ignoreReplicaException(failure)) {
+                            throw failure;
```
thanks for the suggestion! Added asserts for Version conflict and mapping parse failures

areek commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:
```diff
@@ -44,81 +44,61 @@
+                                  @Nullable Location location, @Nullable Exception operationFailure,
+                                  IndexShard primary) {
+            super(request, finalResponse, operationFailure);
+            assert operationFailure != null ^ finalResponse != null;
```
done

areek replied:
we can't asssert on location when operationFailure != null, as we can have noop update operation which will return a null location

bleskes replied:
but still if the location is != null, there must be no failure?

areek commented:
Thanks @bleskes for the feedback. I updated the PR, addressing all your comments, including adding tests for failure handling in `TransportWriteAction` and `InternalEngine`

s1monw commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -386,45 +388,56 @@ private long checkDeletedAndGCed(VersionValue versionValue) {
+        if (isDocumentFailure) {
+            return failure;
+        } else {
+            ElasticsearchException exception = new ElasticsearchException(failure);
```
sorry but this is an absolute no-go. We worked so hard to get rid of this in so many places. We should just rethrow the original exception instead of hiding it in and ElasticsearchException. Yet, I think we can use a littel hack to make it work in this case:
```Java


    @SuppressWarnings("unchecked")
    static <T extends Throwable> void rethrow(Throwable t) throws T {
        throw (T) t;
    }
```

should allow you to just retrhow `failure` without declaring it. I think in this case it's OK

areek replied:
Thanks @s1monw for proposing a solution to this. Now we rethrow the original exception, if it failed the engine instead of wrapping it in ElasticsearchException.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:
```diff
@@ -44,81 +44,61 @@
+            ReplicaRequest extends ReplicatedWriteRequest<ReplicaRequest>,
             Response extends ReplicationResponse & WriteResponse
-        > extends TransportReplicationAction<Request, Request, Response> {
+        > extends TransportReplicationAction<Request, ReplicaRequest, Response> {
```
maybe - we'll see :) I'm fine with having it consistent with TRA

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -376,11 +370,22 @@ public void onFailure(Exception e) {
+         * expects <code>finalResponseIfSuccessful</code> or <code>finalFailure</code> to be not-null
+         */
+        public PrimaryResult(ReplicaRequest replicaRequest, Response finalResponseIfSuccessful, Exception finalFailure) {
+            assert finalFailure != null ^ finalResponseIfSuccessful != null : "either a response or a failure has to be not null";
```
can we add some info to the string message to let us know which whether both were null or both were set?

areek replied:
added `finalFailure` and `finalResponseIfSuccessful` as part of the assert message

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -133,44 +134,49 @@ protected WritePrimaryResult shardOperationOnPrimary(BulkShardRequest request, I
             // execute item request
             final Engine.Result operationResult;
             final DocWriteResponse response;
+            BulkItemRequest replicaRequest = request.items()[requestIndex];
```
wondering whether we should make it final, so each path will have to set it. wdyt?

areek replied:
this makes sense, made `replicaRequest` final and explicitly set it for all the cases

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -133,44 +134,49 @@ protected WritePrimaryResult shardOperationOnPrimary(BulkShardRequest request, I
-                }
-                // set update response
-                item.setPrimaryResponse(new BulkItemResponse(item.id(), opType, response));
+            request.items()[requestIndex] = replicaRequest;
```
hmm... strictly speaking we should restore this on retry, like we do with  preVersions & preVersionTypes. This is an existing bug. can you open an issue about it so we can think about a proper solution in a follow up? This one is hairy enough :)

areek replied:
I opened https://github.com/elastic/elasticsearch/issues/21221 for this, I agree we should restore the original request, but it is non-trivial because we update the request in place. 
In reality though, it should not be a problem, because upon retrying the shard bulk request, the shard will just execute the translated request instead of translating the update request again.

bleskes replied:
not completely true as without the indexing into the replicas, we don't know if the way we resolved the update is valid.

areek replied:
when could a resolved update request that successfully executed on the primary not be valid for a replica?

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:
```diff
@@ -182,20 +188,14 @@ protected WritePrimaryResult shardOperationOnPrimary(BulkShardRequest request, I
                 // then just use the response we got from the successful execution
-                if (item.getPrimaryResponse() == null || isConflictException(failure) == false) {
-                    item.setPrimaryResponse(new BulkItemResponse(item.id(), docWriteRequest.opType(),
+                if (replicaRequest.getPrimaryResponse() == null || isConflictException(failure) == false) {
```
This is a mess - there is no guarantee from the coe that the replica request wasn't just replaced when we did `request.items()[requestIndex] = replicaRequest;` . In practice I think it's OK, but I don't like it. Also, now that we have a proper primary relocation (where we don't shut down the shards half way operations), I don't think it's needed. Can you open another follow up issue so we can evaluate carefully?

areek replied:
I will open a followup issue, for some reason github is erroring out on creating issues now :0

areek replied:
opened https://github.com/elastic/elasticsearch/issues/21230 for this

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -327,44 +334,46 @@ public long getTook() {
+        void setTranslogLocation(Translog.Location translogLocation) {
             if (freeze == false) {
-                this.location = location;
+                this.translogLocation = translogLocation;
```
do we also want to assert failure is null?

areek replied:
added an assert for a null failure

bleskes commented on core/src/main/java/org/elasticsearch/ElasticsearchException.java:
```diff
@@ -580,8 +579,7 @@ public String toString() {
                 org.elasticsearch.action.RoutingMissingException::new, 79),
-        OPERATION_FAILED_ENGINE_EXCEPTION(OperationFailedEngineException.class,
-                OperationFailedEngineException::new, 80),
+        // 80 used to be for IndexFailedEngineException, removed in 6.0
```
we need a bwc here no? or is it somewhere else?

areek replied:
@bleskes we need bwc here if we care about rolling restarts from 5.x. I am not sure how to add bwc for this (i.e. might be missing something). We could simply add the exceptions back (but not use them anywhere) so they can be serialized/deserialized when a 6.0 node is acting as a coordinating node? 

Another question: The index/delete operation failures were communicated as exceptions, so do we even need a bwc for these failures for serialization/deserialization or can we just rely on generic exception serialization/deserialization like we currently do for persistent engine failures during index/delete operations?

bleskes commented on core/src/main/java/org/elasticsearch/indices/IndexingMemoryController.java:
```diff
@@ -201,17 +201,23 @@ protected void deactivateThrottling(IndexShard shard) {
-        statusChecker.bytesWritten(result.getSizeInBytes());
+    private void recordOperationBytes(Engine.Operation operation, Engine.Result result) {
+        final int sizeInBytes;
+        if (result.getTranslogLocation() != null) {
```
I talked this one with @mikemccand and since this whole thing is a heuristic to approximate the memory usage of the IW, we think it's OK to always use the `operation.estimatedSizeInBytes()` . No need to look at the translog location.

bleskes replied:
++ sorry for not mentioning that it was already the case before - I know you just ported what was there 🎉

bleskes commented on core/src/test/java/org/elasticsearch/action/support/replication/TransportWriteActionTests.java:
```diff
@@ -124,15 +126,46 @@ public void testReplicaWaitForRefresh() throws Exception {
+                TestAction.WriteReplicaResult::respond);
+    }
+
+    private <Result, Response> void handleDocumentFailure(TestAction testAction,
```
nit: can you unpack it? it saves on 10 lines of code at the expense of making it much harder to understand what it does - you end going back and forth - I think it's not worth it?

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexingOperationListener.java:
```diff
@@ -40,7 +40,7 @@
+    default void postIndex(Engine.Index index, Engine.IndexResult result) {}
 
     /**
      * Called after the indexing operation occurred with exception.
```
can we document the difference with the other postIndex and how failures are treated?

areek commented:
@bleskes thanks again for the review, I addressed all the comments and had 
one question regarding bwc for the removed exceptions in https://github.com/elastic/elasticsearch/pull/19105#discussion_r85872981

bleskes commented:
@areek because the github ui sucks, I'm responding here so it will be easy to see:

> >  if we care about rolling restarts from 5.x

We care! and yeah, I think we can just keep them in there and doc that they can be removed in 7.0 (assertion?)

> The index/delete operation failures were communicated as exceptions, so do we even need a bwc for these failures for serialization/deserialization or can we just rely on generic exception serialization/deserialization like we currently do for persistent engine failures during index/delete operations?

I  don't think we can rely on generic exceptions for incoming responses from old nodes. That will ask for a specific exception ID and we won't have it -> boom.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:
```diff
@@ -89,6 +89,9 @@ public WritePrimaryResult(ReplicaRequest request, @Nullable Response finalRespon
                                   IndexShard primary) {
             super(request, finalResponse, operationFailure);
+            if (location != null) {
+                assert operationFailure == null : "expected no failures when translog location is not null";
```
nit: assertion location == null || operationFailure == null ?

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexingOperationListener.java:
```diff
@@ -38,12 +38,17 @@
     /**
-     * Called after the indexing operation occurred.
+     * Called after the indexing operation occurred. Implementations should
+     * check {@link Engine.IndexResult#hasFailure()} for operation failures
```
I don't think we need to say this? how about "note that this method is also called when indexing a document didn't succeed because of document related failures. See {@link #postIndex(..)} for engine level failures.

s1monw commented on core/src/main/java/org/elasticsearch/index/engine/DeleteFailedEngineException.java:
```diff
@@ -1,36 +0,0 @@
-
-import java.io.IOException;
-
-public class DeleteFailedEngineException extends EngineException {
```
👍

s1monw commented on core/src/main/java/org/elasticsearch/index/IndexingSlowLog.java:
```diff
@@ -133,23 +133,20 @@ private void setReformat(boolean reformat) {
-    }
-
-
-    private void postIndexing(ParsedDocument doc, long tookInNanos) {
```
👍

s1monw commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -277,9 +278,119 @@ public Condition newCondition() {
+    public abstract IndexResult index(Index operation);
 
-    public abstract void delete(Delete delete) throws EngineException;
+    public abstract DeleteResult delete(Delete delete);
```
a java doc commetn would be great here while we are at it

s1monw commented on core/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -277,9 +278,119 @@ public Condition newCondition() {
+        private final Exception failure;
+        private Translog.Location translogLocation;
+        private long took;
+        private boolean freeze;
```
maybe use `private final SetOnce<Boolean> frozen` here it will barf if you freeze multiple times?

s1monw commented on core/src/main/java/org/elasticsearch/index/engine/IndexFailedEngineException.java:
```diff
@@ -1,63 +0,0 @@
-import java.io.IOException;
-import java.util.Objects;
-
-public class IndexFailedEngineException extends EngineException {
```
👍

s1monw commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -387,46 +388,60 @@ private long checkDeletedAndGCed(VersionValue versionValue) {
+            // and set the error in operation.setFailure. In case of environment related errors, the failure
+            // is bubbled up
+            isDocumentFailure = !((failure instanceof IllegalStateException || failure instanceof IOException)
+                    && maybeFailEngine(operation.operationType().getLowercase(), failure));
```
can we please move the `maybeFailEngine(operation.operationType().getLowercase(), failure))` one line above and assign it to a variable. I really wanna make sure no short circuit logic here prevents it from being called

s1monw commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -387,46 +388,60 @@ private long checkDeletedAndGCed(VersionValue versionValue) {
+
+    @SuppressWarnings("unchecked")
+    static <T extends Throwable> void rethrow(Throwable t) throws T {
+        throw (T) t;
```
maybe document that this is a hack to retrhow the original

s1monw commented on core/src/main/java/org/elasticsearch/index/shard/InternalIndexingStats.java:
```diff
@@ -74,14 +74,18 @@ IndexingStats stats(boolean isThrottled, long currentThrottleInMillis, String...
+        if (result.hasFailure() == false) {
+            if (!index.origin().isRecovery()) {
+                long took = result.getTook();
+                totalStats.indexMetric.inc(took);
```
should we have a write failures statistic too? @bleskes
